### PR TITLE
[Debt] Bump `react-hook-form`

### LINKED ISSUE #5282: Bump react-hook-form to >= 7.41
Seems like there are some breaking changes in react-hook-form.  Upgrade the dependency to the latest version (at least 7.41) and make any needed changes.

Acceptance Criteria
- [ ] Bump react-hook-form to >= 7.41
- [ ] App continues to function as normal

### PR DESCRIPTION
**Pull request by esizer**
🤖 Resolves #5282 

## 👋 Introduction

This bumps `react-hook-form` and massages some types to get it to work properly.

## 🕵️ Details

This has been sitting for a while as we wait for `react-hook-form` to make some types external. However, I was able to get it to work by massaging some types. It is a less than ideal solution though, the only place this was done was in `UnsavedChanges` which is currently a pain point anyway so I'm not sure how much negative impact will result from these changes 🤷‍♀️ 

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Install deps `npm i`
2. Build app `npm run build` and confirm it succeeds without error
3. Navigate to anywhere we are using the `BasicForm` (profile is a good place for this)
4. Interact with the form confirming it works as expected
    - Can submit
    - Error summary appears
    - Unsaved changes summary appears
    - No console errors

### CODE REVIEW

File: packages/forms/src/components/BasicForm.tsx [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import React, { PropsWithChildren, ReactElement } from "react";
 import {
+  DeepPartial,
```
Comment (mnigh) [inline packages/forms/src/components/BasicForm.tsx:3]: `DeepPartial` is defined but never used

File: packages/forms/src/components/BasicForm.tsx [hunk 2]
```diff
@@ -108,14 +108,18 @@ function BasicForm<TFieldValues extends FieldValues>({
         /**
          * Iterates through all cached values touching and dirtying the fields
          * so we can display any errors and unsaved changes
+         *
+         * Note: There is a lot of type hacking with `as` here since
+         * `react-hook-form` does not export the required types
```
Comment (mnigh) [inline packages/forms/src/components/BasicForm.tsx:113]: Worth opening an issue with the library?

✍️ https://github.com/react-hook-form/react-hook-form/issues/new?assignees=&labels=feature+request%2C+waiting-up-vote&template=feature_request.md&title=

**Comment by vd1992** [comment]
ExperienceForm.test.tsx -> "should submit with required fields"
Consistently fails locally too.

File: packages/forms/src/components/BasicForm.tsx (hunk 2 above)
Comment (esizer) [reply packages/forms/src/components/BasicForm.tsx:113]: I'm not sure, there is no real reason to need that type. We are just doing something very funky to render those cached values that doesn't really make sense for the library. We could try though, who knows 🤷‍♀️

**Comment by esizer** [comment]
> ExperienceForm.test.tsx -> "should submit with required fields"
> Consistently fails locally too.

Okay, should be fixed as part of [fix experience form test](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6180/commits/708fc1c8c74296ff7bd174afafe7abc623cb9563). I think what was happening was `react-hook-form` is all async so for some reason validation was not working. I updated it to manually update the form instead of pre-populating and praying.

[fix experience form test](https://github.com/GCTC-NTGC/gc-digital-talent/pull/6180/commits/708fc1c8c74296ff7bd174afafe7abc623cb9563)

**Review by vd1992 — APPROVED** [review]
Can't really comment on your "hack" but...
Jest ✅ 
E2E ✅ 
Chromatic ✅ 

Gave it a quick whirl locally and seems good. Nice! Good to get this one out of the way.

**Pull request merged**